tests/std/tests/P2231R1_complete_constexpr_optional_variant/test.cpp

```diff
@@ -0,0 +1,345 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <assert.h>
+#include <initializer_list>
+#include <optional>
+#include <type_traits>
+#include <utility>
+#include <variant>
+
+using namespace std;
+
+struct With_trivial_destructor {
+    int _val = 0;
+    constexpr With_trivial_destructor(const int val) noexcept : _val(val) {}
+    constexpr With_trivial_destructor(initializer_list<int> vals) noexcept : _val(*vals.begin()) {}
+
+    constexpr bool operator==(const int right) const noexcept {
+        return _val == right;
+    }
+};
+
+struct With_nontrivial_destructor {
+    int _val = 0;
+    constexpr With_nontrivial_destructor(const int val) noexcept : _val(val) {}
+    constexpr With_nontrivial_destructor(initializer_list<int> vals) noexcept : _val(*vals.begin()) {}
+    constexpr ~With_nontrivial_destructor() {}
+
+    constexpr bool operator==(const int right) const noexcept {
+        return _val == right;
+    }
+};
+
+struct Dummy {
+    int _val = 0;
+};
+
+template <class T>
+constexpr bool test_optional() {
+    { // empty construction
+        optional<T> default_constructed;
+        assert(!default_constructed.has_value());
+
+        optional<T> nullopt_constructed{nullopt};
+        assert(!nullopt_constructed.has_value());
+    }
+
+    { // construction from underlying type
+        const T input{42};
+        optional<T> copy_constructed{input};
+        assert(copy_constructed.has_value());
+        assert(*copy_constructed == 42);
+
+        optional<T> move_constructed{T{42}};
+        assert(move_constructed.has_value());
+        assert(*move_constructed == 42);
+
+        optional<T> copy_assigned;
+        assert(!copy_assigned.has_value());
+        copy_assigned = input;

```

Comment:
The names of the objects on 50 an 54 refer to the operation we perform on the contained value, but 58 and 64 do not since 60 and 66 actually construct the contained value rather than assign to it. Are these names simply copy pasta, and if so, could we change them to something less confusing like `o1`, `o2`, ... ?

Reply:
Renamed those to `construct_via_copy_assignment` / `construct_via_move_assignment`

I really try to give the opbjects a descriptive name of what happens to ensure that I do not miss a case, which is much easier when naming things `oN`

Reply:
Thank you! Naming is hard, but this is a huge improvement - I appreciate it.